[ENH] HMM annotation estimator

miraep8 opened the pull request:
I wanted to share a draft PR of what I have been working on with HMM estimator. 


<details>
<summary>More Details:</summary>
<br>

- rather than assuming a particular emission distribution function, which I believe is what `hmmlearn` does (ie you can choose to use a Guassian HMM vs a Multinomial Discrete HMM, and you need to either build, or hope an HMM with your emission distribution has already been built), we simply accept a list of functions which behave as pdf and use those to calculate the emission probability.  The viterbi algorithm gets to treat those as a black box. 
    - the hope here is that we can add some functionality which automatically generates some nice functions for people if needed (ie give me an n-dimensional gaussian!), but it also gives the flexibility to do more interesting things!  - Have an empirical distribution you want to act as a "pdf"?  Have some other distribution you are getting from some other system that you would like to use? As long as it can return a valid probability then it *should* work with our HMM estimator, no need to write your own!
- I have spent some time optimizing the algorithms with numpy (in particular, the `_calculate_trans_mats` function which is the slowest part), the plan is to push this a bit more and then conduct speed tests between this draft estimator and `hmmlearn`, so we can have a more quantitative comparison of speed. 
- I implemented a very simple version of a base object, named `SimpleBaseEstimator` and used that here.  It is very similar to what Lukasz did in his annotation flavored PR actually, just added in the `get_params()` and `set_params()` functions, which are technically required for it to be a sklearn like estimator. 
    - I am not proposing we replace our Base Object with what I wrote at all, just wanted to see how many of the bells and whistles in Base Estimator I could get away without! Again - just for testing purposes.


### Example of how to use it: 

```
from sktime.annotation.hmm import HMM
from scipy.stats import norm
from numpy import asarray

centers = [3.5,-5]
sd = [.25 for i in centers]
emi_funcs = [(norm.pdf, {'loc': mean, 'scale': sd[ind]}) for ind, mean in enumerate(centers)]

test = HMM(emi_funcs, asarray([[0.25,0.75], [0.666, 0.333]]))
obs = asarray([3.7,3.2,3.4,3.6,-5.1,-5.2,-4.9])
test.fit(obs)
print(test.predict())
```

<\details>

@fkiraly, @lmmentel, @GuzalBulatova

fkiraly commented:
>the hope here is that we can add some functionality which automatically generates some nice functions for people if needed

there currently is a dormant project which meant to introduce distribution interfaces to the ML context:
https://github.com/alan-turing-institute/skpro
maybe we can revisit that sometime. E.g., have `BaseDistribution` inherit from `BaseObject` 😄

fkiraly commented:
>I am not proposing we replace our Base Object with what I wrote at all, just wanted to see how many of the bells and whistles in Base Estimator I could get away without! Again - just for testing purposes.

hmmmmm..... all this denial just makes it more suspicious :pouting_cat:

miraep8 commented:
Also - I **think** unrelated to this PR, there seem to have been some strange failures on the last CI/CD runthrough.  Can  see details by clicking on the small red x next to the failed run, but copied the error message below in case it is a more general issue, note the error only cropped up in `Install and test / test-mac (3.7) (pull_request)` and the other os/python versions seemed to work ok....  Probably just a one off, but flagging this just in case. 

```
=================================== FAILURES ===================================
____ [doctest] sktime.datasets._single_problem_loaders.load_UCR_UEA_dataset ____
[gw1] darwin -- Python 3.7.13 /Users/runner/hostedtoolcache/Python/3.7.13/x64/bin/python
088         series associated with each case. If return_X_y is False, column
089         n_dimensions+1 contains the class labels/target variable.
090     y: numpy array, optional
091         The class labels for each case in X, returned separately if return_X_y is
092         True, or appended to X if False
093 
094     Examples
095     --------
096     >>> from sktime.datasets import load_UCR_UEA_dataset
097     >>> X, y = load_UCR_UEA_dataset(name="Yoga")
UNEXPECTED EXCEPTION: ContentTooShortError('retrieval incomplete: got only 65422 out of 20908976 bytes')
Traceback (most recent call last):
  File "/Users/runner/hostedtoolcache/Python/3.7.13/x64/lib/python3.7/doctest.py", line 1337, in __run
    compileflags, 1), test.globs)
  File "<doctest sktime.datasets._single_problem_loaders.load_UCR_UEA_dataset[1]>", line 1, in <module>
  File "/Users/runner/hostedtoolcache/Python/3.7.13/x64/lib/python3.7/site-packages/sktime/datasets/_single_problem_loaders.py", line 99, in load_UCR_UEA_dataset
    return _load_dataset(name, split, return_X_y, extract_path)
  File "/Users/runner/hostedtoolcache/Python/3.7.13/x64/lib/python3.7/site-packages/sktime/datasets/_data_io.py", line 147, in _load_dataset
    extract_path=extract_path,
  File "/Users/runner/hostedtoolcache/Python/3.7.13/x64/lib/python3.7/site-packages/sktime/datasets/_data_io.py", line 73, in _download_and_extract
    urlretrieve(url, zip_file_name)
  File "/Users/runner/hostedtoolcache/Python/3.7.13/x64/lib/python3.7/urllib/request.py", line 288, in urlretrieve
    % (read, size), result)
urllib.error.ContentTooShortError: <urlopen error retrieval incomplete: got only 65422 out of 20908976 bytes>
/Users/runner/hostedtoolcache/Python/3.7.13/x64/lib/python3.7/site-packages/sktime/datasets/_single_problem_loaders.py:97: UnexpectedException
```

fkiraly commented:
@miraep8, clearly, your content is too short. Please fix the issue by making your content longer.

miraep8 commented:
By the way @lmmentel, not saying that the test I am using is great (it is in fact not), but have a snippet of code in the testing file for generating a gaussian with shifting mean.  Plan to make this better (ie - just generate a file and load that as well instead of generating the data in real time) and also add some other tests.

miraep8 commented:
Related to my issue #2799

miraep8 commented:
Testing update, have been comparing with the Gaussian estimator from `hmmlearn`.  It seems like their package is much faster than mine, about 20-50 times after it seems, but that (at least in the synthetic data I have been running it on so far) ours seems to get closer to the truth! 

Here is code I have run to test it on!

```
#imports
from sktime.annotation.hmm import HMM
from scipy.stats import norm
import numpy as np
from hmmlearn.hmm import GaussianHMM
import time

#generating fake data: 
centers = [3.5,-5, 8, 10, .5]
sd = [.25 for i in centers]
seg_lengths = [1000, 500, 250 , 700, 150]
gauss_data = np.zeros([sum(seg_lengths)])
labels = np.zeros([sum(seg_lengths)])
i = 0
for ind, s_len in enumerate(seg_lengths):
    gauss_data[i:i+s_len] = np.random.default_rng().normal(loc=centers[ind], scale=sd[ind], size=[s_len])
    labels[i:i+s_len] = np.asarray([ind for _ in range(s_len)])
    i+= s_len

#running comparison:   
def convert_hmmlearn_output(old_array):
    output_dict = dict()
    count = 0
    new_array = []
    for i in old_array:
        if not i in output_dict.keys():
            output_dict[i] = count
            count += 1
        new_array.append(output_dict[i])
    return new_array

## Time for sktime HMM estimator:
start_time = time.time()
emi_funcs = [(norm.pdf, {'loc': mean, 'scale': 1}) for mean in centers]
transition_matrix = np.asarray([[((1/(s_len/14))/(len(centers)-1)) for state in centers] for s_len in seg_lengths])
for ind, s_len in enumerate(seg_lengths):
    transition_matrix[ind, ind] = 1 - (1/(s_len/14))
sktime_hmm = HMM(emi_funcs, transition_matrix)
sktime_hmm.fit(gauss_data)
sktime_predicted_labels = sktime_hmm.predict()
sktime_time = time.time() - start_time
print(f'sktime HMM estimator took {sktime_time:.4f} seconds to run')

## Time for hmmlearn HMM estimator:
start_time = time.time()
transition_matrix = np.asarray([[((1/(s_len/14))/(len(centers)-1)) for state in centers] for s_len in seg_lengths])
hmmlearn_hmm = GaussianHMM(n_components=len(centers), means_prior = np.asarray(centers).reshape(-1, 1), covars_prior = sd, covariance_type='spherical',
                           transmat_prior = transition_matrix)
hmmlearn_hmm.fit(gauss_data.reshape(-1, 1))
hmmlearn_predicted_labels = hmmlearn_hmm.predict(gauss_data.reshape(-1, 1))
hmmlearn_time = time.time() - start_time
print(f'hmmlearn HMM estimator took {hmmlearn_time:.4f} seconds to run')

if sktime_time > hmmlearn_time:
    print(f'Current sktime implementaiton is {(sktime_time/hmmlearn_time):.1f} time slower than hmmlearn')
else:
    print(f'Current sktime implementaiton is {(hmmlearn_time/sktime_time):.1f} time slower than hmmlearn')
hmmlearn_predicted_labels =convert_hmmlearn_output(hmmlearn_predicted_labels)
print(f'The percent agreement between the two algorithms {100*sum(hmmlearn_predicted_labels==sktime_predicted_labels)/len(gauss_data): .1f}%')
print(f'sktime agreement with ground truth: {100*sum(sktime_predicted_labels == labels)/len(labels): .1f}%')
print(f'hmmlearn agreement with ground truth: {100*sum(hmmlearn_predicted_labels == labels)/len(labels): .1f}%')

```

Which prints out: 

```
sktime HMM estimator took 0.8916 seconds to run
hmmlearn HMM estimator took 0.0371 seconds to run
Current sktime implementaiton is 24.0 time slower than hmmlearn
The percent agreement between the two algorithms  84.7%
sktime agreement with ground truth:  100.0%
hmmlearn agreement with ground truth:  84.7%
```

This is just one pass through, so will be collecting more samples, but thought this was an interesting observation.

miraep8 commented:
@lmmentel @KatieBuc @fkiraly - I have refactored my HMM estimator to make use of the current BaseAnnotation class.  The current plan is to get it merged in its currently compatible format, and move some of the BaseEstimator stuff I was originally exploring here into a different PR. I also plan to add an option - 'use_hmmlearn' as an option to this estimator in a future PR. Ready for review! :)

fkiraly commented:
>I also plan to add an option - use_hmmlearn as an option to this estimator in a future PR

I would do that in a separate class - we have had similar discussions earlier on the multiple ARIMA models (I cannot find the issue anymore), on whether they should be in a single class or in multiple. The consensus was, they should be multiple, as long as it is coming from different packages with different interfaces - to isolate dependencies, and to follow the conceptual model of "one estimator = one parametric algorithm" more clearly.

fkiraly commented on sktime/annotation/hmm.py:
```diff
@@ -0,0 +1,275 @@
+class HMM(BaseSeriesAnnotator):
+    """Implements a simple HMM fitted with viterbi algorithm.
+
+    The HMM annotation estimator facilitates the appilcation of
+    the vertibi algorithm to fit a sequence of 'hidden state' class
```
Viterbi

lmmentel replied:
alternative wording 

```suggestion
    The HMM annotation estimator uses
    the Veterbi algorithm to fit a sequence of 'hidden state' class
```

fkiraly commented on sktime/annotation/hmm.py:
```diff
@@ -0,0 +1,275 @@
+    annotations (represented by an array of integers the same size
+    as the observation) to a sequence of observations.
+
+    This is done by finding the most likely path given the emission
```
It would be crucial to be more precise to describe what the output is, given input.
I.e., given the X, what is being produced in `predict`? What is the fit doing?

The current docstring is just a high-level description of HMM, not of the specific procedure that fit/predict below carries out.

miraep8 replied:
I added some more details to that effect (and also moved all the logic to predict as you had suggested). Are you happier with it now, or does it still need more?

fkiraly commented on sktime/annotation/hmm.py:
```diff
@@ -0,0 +1,275 @@
+
+    Parameters
+    ----------
+    emission_funcs : a list of functions, which should be properly
```
must be more precise. What signature do the functions have?
Should also probably be callable.

fkiraly commented on sktime/annotation/hmm.py:
```diff
@@ -0,0 +1,275 @@
+    ----------
+    emission_funcs : a list of functions, which should be properly
+        normalized PDFs over the same space as the observed data.
+    transition_prob_mat: a nxn array of probabilities, where 'n'
```
please provide python type

fkiraly commented on sktime/annotation/hmm.py:
```diff
@@ -0,0 +1,275 @@
+        row should sumn to 1 in order to be properly normalized
+        (ie the j'th column in the i'th row represents the
+        probability of transitioning from state i to state j.)
+    initial_probs: optional, a 1d array of probabilities across
```
please specify python types, and specify which dimensions must be equal, e.g., should the length match something else, above?

fkiraly commented on sktime/annotation/hmm.py:
```diff
@@ -0,0 +1,275 @@
+    >>> sd = [.25 for i in centers]
+    >>> emi_funcs = [(norm.pdf, {'loc': mean,
+    ...  'scale': sd[ind]}) for ind, mean in enumerate(centers)]
+    >>> test = HMM(emi_funcs, asarray([[0.25,0.75], [0.666, 0.333]]))
```
re naming, avoid calling things `test` or `temp` or similar. Names, even if short, should be as descriptive as possible. Better, perhaps: `hmm`, `hmm_est`

fkiraly commented on sktime/annotation/hmm.py:
```diff
@@ -0,0 +1,275 @@
+        self.emission_funcs = emission_funcs
+        self.transition_prob_mat = transition_prob_mat
+        self.states = [i for i in range(self.num_states)]
+        self._validate_init()
```
should happen after `super.__init__` call

fkiraly commented on sktime/annotation/hmm.py:
```diff
@@ -0,0 +1,275 @@
+
+    def _validate_init(self):
+        """Verify that the parameters passed to init are well behaved."""
+        tran_mat_len = len(self.transition_prob_mat[:, 0])
```
clearer: `self.transition_prob_mat.shape[0]`

miraep8 replied:
Yes, good suggestion!

fkiraly commented on sktime/annotation/hmm.py:
```diff
@@ -0,0 +1,275 @@
+        """Verify that the parameters passed to init are well behaved."""
+        tran_mat_len = len(self.transition_prob_mat[:, 0])
+        # transition_prob_mat should be square:
+        if not len(self.transition_prob_mat.shape) == 2 or not tran_mat_len == len(
```
use `x.ndim` instead of `len(x.shape)`

fkiraly commented on sktime/annotation/hmm.py:
```diff
@@ -0,0 +1,275 @@
+            init_prob_len = len(self.initial_probs)
+        if not tran_mat_len == self.num_states == init_prob_len:
+            raise ValueError(
+                "Number of hidden states is inconsistent!  Was passed "
```
say more explicitly:
* what *needs* to be the case
* which variables are being referred to here, so the user can compare to the function docstring by matching variable names rather than trying to match text

fkiraly commented on sktime/annotation/hmm.py:
```diff
@@ -0,0 +1,275 @@
+        num_obs: int,
+        num_states: int,
+    ) -> Tuple[np.array, np.array]:
+        """Calculate the transition mats used in the viterbi algorithm.
```
capitalize Viterbi (name of a person)

fkiraly commented on sktime/annotation/hmm.py:
```diff
@@ -0,0 +1,275 @@
+
+        Returns
+        -------
+            -hmm_fit: a 2xn array with the first column being position and the second
```
typo?

miraep8 replied:
Yes, also implemented Viterbi in another situation, and the comment was more relevant there. (biological annotation)

fkiraly commented on sktime/annotation/hmm.py:
```diff
@@ -0,0 +1,275 @@
+
+    def _predict(self, X):
+        """Use the transition matrices to calculate the most likely state."""
+        return self._hmm_viterbi_label()
```
this can´t be right - `_hmm_viterbi_label` not using `X`? Can this only be used in-sample?

miraep8 replied:
Well, as you point out below the estimator (as it is) is only applying Viterbi to a single observation (and originally all the logic that extracted information from `X` was in `_fit`.  I have since moved everything to `_predict`.

fkiraly reviewed (changes requested):
I think this is just the Viterbi algorithm that infers hidden labels based on known initial and transition probas, not Baum-Welch, correct?
In that case, I would not implement a `_fit`, or check in `_ predict` whether the `_fit` labels have already been seen. 
The `_predict` should always produce the inferred hidden states for its argument `X`,  but currently it produces those for `X`.

The more "interesting" version of thie would use Baum-Welch or another form of EM in `fit`, and `Viterbi` in `predict`.

lmmentel commented on sktime/annotation/hmm.py:
```diff
@@ -0,0 +1,275 @@
+HMM Annotation Estimator.
+
+Implements a basic Hidden Markov Model (HMM) as an annotation estimator.
+To read more about the algorithm, check out the wikipedia page:
+(https://en.wikipedia.org/wiki/Hidden_Markov_model
```
how about?
```suggestion
To read more about the algorithm, check out the `HMM wikipedia page <https://en.wikipedia.org/wiki/Hidden_Markov_model>`_.
```

lmmentel commented on sktime/annotation/hmm.py:
```diff
@@ -0,0 +1,275 @@
+        self.num_states = len(emission_funcs)
+        self.emission_funcs = emission_funcs
+        self.transition_prob_mat = transition_prob_mat
+        self.states = [i for i in range(self.num_states)]
```
all attributes that are not part of the `__init__` signature should be assigned in `fit` to avoid side effects when using `set_params`

lmmentel commented on sktime/annotation/hmm.py:
```diff
@@ -0,0 +1,275 @@
+                f" as the size of the transition matrix, and  "
+                f" {init_prob_len} number of initial state probs"
+            )
+        for i in range(len(self.transition_prob_mat)):
```
how about using `np.isclose` and compare against a vector of ones to avoid the loop?

miraep8 replied:
Yes, that is cleaner

lmmentel commented on sktime/annotation/tests/test_all_annotators.py:
```diff
@@ -23,4 +24,4 @@ def test_output_type(Estimator):
     args = _make_args(estimator, "predict")
     y_pred = estimator.predict(*args)
-    assert isinstance(y_pred, pd.Series)
+    assert isinstance(y_pred, pd.Series) or isinstance(y_pred, np.ndarray)
```
this could work as well

```suggestion
    assert isinstance(y_pred, (pd.Series, np.ndarray))
```

miraep8 commented:
Thanks both @lmmentel and @fkiraly for the useful feedback!  I have implemented your suggestions.

Regarding Baum-Welch/other EM - I agree that would definitely be more interesting.  I think my original plan was to add that on in a future PR, if that is ok with you? 

However I have now moved all logic out of `_fit` as was suggested and everything has been consolidated in `_predict`

GuzalBulatova commented on sktime/annotation/hmm.py:
```diff
@@ -0,0 +1,314 @@
+
+    Parameters
+    ----------
+    emission_funcs : a list of length n.  Should be either a list of
```
My suggestions for improvement regarding the docstrings: an example of conventional docs can be found in [boss.py file](https://github.com/alan-turing-institute/sktime/blob/main/sktime/classification/dictionary_based/_boss.py). For example, first line is the type, and broader description of a parameter starts from the next line, which makes it further easier to read. Sth like
```
parameter : float, optional, default=1.5
   Broader description of the parameter.
```

miraep8 replied:
Thank you!  It was useful to see a more concrete example of what a good example of docstring looks like :)

GuzalBulatova replied:
This looks great, thanks for the changes! 
I think the only docstr left is `_calculate_trans_mats`, and it'll be good to go.

miraep8 replied:
Ah yes, good catch!  Will work on that now.  Thanks again for the helpful feedback! :)

GuzalBulatova commented on sktime/annotation/hmm.py:
```diff
@@ -0,0 +1,314 @@
+        max_inds = np.zeros(self.num_obs, dtype=np.int32)
+        max_inds[-1] = np.argmax(self.trans_prob[:, -1])
+        hmm_fit[-1] = self.states[max_inds[-1]]
+        for index in reversed(list(range(1, self.num_obs))):
```
This might read less intuitive but we could use just `range` with all its params `(start, stop, step)` here, like 
`for index in range(self.num_obs, 0, -1)`, 
what do you think?

miraep8 replied:
I think that reads a lot better actually!  Thanks for the suggestion!

GuzalBulatova reviewed (commented):
Without exaggeration, this is a great contribution. Amazing work, well done, Mirae! 
I personally am especially grateful for type hints in methods, we should (or maybe already have started to) make it a common practice in sktime :)
I only have some minor style suggestions

fkiraly reviewed (approved):
I think this is ok, great!

Re docstring, I think this is still a bit confusing - lot of text, but doesn't tell the user precisely. Not a blocker, but let's discuss at some later point.

lmmentel commented:
@fkiraly I cannot merge since the button is inactive for me since CI is waiting for the new checks. I'm happy for this to be merged.
